File: 1. Two Sum/Main.java
import java.util.HashMap;
import java.util.Map;

class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        var ans = solution.twoSum(new int[]{3, 3}, 6);
        System.out.println(ans[0] + " - " + ans[1]);
    }
}

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            int diff = target - nums[i];
            
            if (map.containsKey(diff)) {
                return new int[] {map.get(diff), i};
            }
            
            map.put(nums[i], i);
        }

        return new int[]{};
    }
}

File: 1. Two Sum/main.go
package main

import "fmt"

func main() {
	fmt.Println(twoSum([]int{3, 3}, 6))
}

func twoSum(nums []int, target int) []int {
	// map[expected]index
	m := make(map[int]int)

	for i := 0; i < len(nums); i++ {
		if j, ok := m[nums[i]]; ok {
			return []int{j, i}
		}
		m[target-nums[i]] = i
	}

	return []int{}
}
